ledger/acctupdates.go

```diff
@@ -853,11 +886,118 @@ func (au *accountUpdates) accountsInitialize(tx *sql.Tx) (basics.Round, error) {
 			}
 			accountIdx += trieRebuildAccountChunkSize
 		}
+
+		// we've just updated the markle trie, update the hashRound to reflect that.
+		err = updateAccountsRound(tx, rnd, rnd)
+		if err != nil {
+			return 0, fmt.Errorf("accountsInitialize was unable to update the account round to %d: %v", rnd, err)
+		}
 	}
 	au.balancesTrie = trie
 	return rnd, nil
 }
 
+// upgradeDatabaseSchema0 upgrades the database schema from version 0 to version 1
+func (au *accountUpdates) upgradeDatabaseSchema0(ctx context.Context, tx *sql.Tx) (updatedDBVersion int32, err error) {
+	au.log.Infof("accountsInitialize initializing schema")
+	err = accountsInit(tx, au.initAccounts, au.initProto)
+	if err != nil {
+		return 0, fmt.Errorf("accountsInitialize unable to initialize schema : %v", err)
+	}
+	_, err = db.SetUserVersion(ctx, tx, 1)
+	if err != nil {
+		return 0, fmt.Errorf("accountsInitialize unable to update database schema version from 0 to 1: %v", err)
+	}
+	return 1, nil
+}
+
+// upgradeDatabaseSchema1 upgrades the database schema from version 1 to version 2
+func (au *accountUpdates) upgradeDatabaseSchema1(ctx context.Context, tx *sql.Tx) (updatedDBVersion int32, err error) {
+	// update accounts encoding.
+	au.log.Infof("accountsInitialize verifying accounts data encoding")
+	modifiedAccounts, err := reencodeAccounts(ctx, tx)
+	if err != nil {
+		return 0, err
+	}
+
+	if modifiedAccounts > 0 {
+		au.log.Infof("accountsInitialize reencoded %d accounts", modifiedAccounts)
+
+		au.log.Infof("accountsInitialize resetting account hashes")
+		// reset the merkle trie
+		err = resetAccountHashes(tx)
+		if err != nil {
+			return 0, fmt.Errorf("accountsInitialize unable to reset account hashes : %v", err)
+		}
+
+		au.log.Infof("accountsInitialize preparing queries")
+		// initialize a new accountsq with the incoming transaction.
+		accountsq, err := accountsDbInit(tx, tx)
+		if err != nil {
+			return 0, fmt.Errorf("accountsInitialize unable to prepare queries : %v", err)
+		}
+
+		// close the prepared statements when we're done with them.
+		defer accountsq.close()
+
+		au.log.Infof("accountsInitialize resetting prior catchpoints")
+		// delete the last catchpoint label if we have any.
+		_, err = accountsq.writeCatchpointStateString(ctx, catchpointStateLastCatchpoint, "")
+		if err != nil {
+			return 0, fmt.Errorf("accountsInitialize unable to clear prior catchpoint : %v", err)
+		}
+
+		au.log.Infof("accountsInitialize deleting stored catchpoints")
+		// delete catchpoints.
+		err = au.deleteStoredCatchpoints(ctx, accountsq)
+		if err != nil {
+			return 0, fmt.Errorf("accountsInitialize unable to delete stored catchpoints : %v", err)
+		}
+	} else {
+		au.log.Infof("accountsInitialize found that no accounts needed to be reencoded")
```

Comment:
Why is this okay? If the database is upgraded from 1 to 2, shouldn't necessarily have accounts reencoded? 
Unless there are no accounts, which can be checked.

Reply:
no. only certain types of accounts needs to be reencoded. If no reencoding was needed, the entire version 1->2 upgrade is not doing anything.
I believe you'll find that this is the case with Betanet ( but not just ).